From 066d96a09e7a969f2cab84a33b74ecc1278f934c Mon Sep 17 00:00:00 2001
From: Tiezhu Yang <yangtiezhu@loongson.cn>
Date: Tue, 17 Dec 2024 09:09:00 +0800
Subject: [PATCH 100/340] FROMLIST: objtool: Handle unreachable entry of rodata

When compiling with Clang on LoongArch, there exists unreachable entry of
rodata which points to a position after the function return instruction,
this is generated by compiler to fill the non-existent switch case, just
skip the entry when parsing the relocation section of rodata.

Signed-off-by: Tiezhu Yang <yangtiezhu@loongson.cn>

Link: https://lore.kernel.org/all/20241217010905.13054-5-yangtiezhu@loongson.cn/
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 tools/objtool/check.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/tools/objtool/check.c b/tools/objtool/check.c
index a38ff31597ec..27be92f344f4 100644
--- a/tools/objtool/check.c
+++ b/tools/objtool/check.c
@@ -1986,9 +1986,10 @@ static int add_jump_table(struct objtool_file *file, struct instruction *insn,
 		if (!dest_insn)
 			break;
 
-		/* Make sure the destination is in the same function: */
-		if (!insn_func(dest_insn) || insn_func(dest_insn)->pfunc != pfunc)
-			break;
+		if (!insn_func(dest_insn) || insn_func(dest_insn)->pfunc != pfunc) {
+			prev_offset = reloc_offset(reloc);
+			continue;
+		}
 
 		alt = malloc(sizeof(*alt));
 		if (!alt) {
-- 
2.48.1

